### PR TITLE
add request metrics

### PR DESCRIPTION
**Pull request by loeffert**
This PR will add metrics to all exposed services / APIs

### CODE REVIEW

File: status/metrics.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		ServiceRequestDuration: serviceRequestDuration("api", namespace),
+	}
+}
+
+func serviceRequestCounter(subsystem, namespace string) *prometheus.CounterVec {
+	vec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "count_service_requests",
```
Comment (smartinov) [inline status/metrics.go]: I'd suggest namespace somethign like foomo and subsystem contentserver

File: server/handlerequest.go [hunk 2]
```diff
@@ -25,37 +32,46 @@ func handleRequest(r *repo.Repo, handler Handler, jsonBytes []byte) (replyBytes
...
 		repoRequest := &requests.Repo{}
 		processIfJSONIsOk(json.Unmarshal(jsonBytes, &repoRequest), func() {
 			reply = r.GetRepo()
 		})
+		addMetrics(metrics, HandlerGetRepo, start, jsonErr, apiErr)
 	default:
 		err = errors.New(log.Error("  can not handle this one " + handler))
 		errorResponse := responses.NewError(1, "unknown handler")
 		reply = errorResponse
+		addMetrics(metrics, "default", start, jsonErr, apiErr)
```
Comment (smartinov) [inline server/handlerequest.go]: Why don't you set metrics after the case? It's always the same including the "handler" from the switch

File: server/handlerequest.go [hunk 3]
```diff
@@ -4,19 +4,26 @@ import (
...
 	"github.com/foomo/contentserver/log"
 	"github.com/foomo/contentserver/repo"
 	"github.com/foomo/contentserver/requests"
 	"github.com/foomo/contentserver/responses"
+	"github.com/foomo/contentserver/status"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
-func handleRequest(r *repo.Repo, handler Handler, jsonBytes []byte) (replyBytes []byte, err error) {
+func handleRequest(r *repo.Repo, handler Handler, jsonBytes []byte, metrics *status.Metrics) (replyBytes []byte, err error) {
```
Comment (smartinov) [inline server/handlerequest.go]: I'd actually suggest using a standard global variable 
Like on their example https://github.com/prometheus/haproxy_exporter/blob/master/haproxy_exporter.go

Or

File: server/handlerequest.go [hunk 4]
```diff
@@ -66,9 +82,30 @@ func handleRequest(r *repo.Repo, handler Handler, jsonBytes []byte) (replyBytes
 		err = apiErr
 		reply = responses.NewError(3, "internal error "+apiErr.Error())
 	}
+
 	return encodeReply(reply)
 }
 
+func addMetrics(metrics *status.Metrics, handlerName Handler, start time.Time, errJSON error, errAPI error) {
```
Comment (smartinov) [inline server/handlerequest.go]: I wouldn't pass in metrics, but would rather suggest having them as global variables

File: status/metrics.go (hunk 1 above)
Comment (smartinov) [reply status/metrics.go]: Also name of the request counter would be service_request_count

File: status/metrics.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		}, []string{MetricLabelHandler, MetricLabelStatus})
+	prometheus.MustRegister(vec)
+	return vec
+}
+
+func serviceRequestDuration(subsystem, namespace string) *prometheus.SummaryVec {
+	vec := prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "time_nanoseconds",
```
Comment (smartinov) [inline status/metrics.go]: Prometheus durations are suggested to be in seconds, and naming should be like service_request_duration_seconds

File: server/handlerequest.go [hunk 6]
```diff
@@ -66,9 +82,30 @@ func handleRequest(r *repo.Repo, handler Handler, jsonBytes []byte) (replyBytes
...
+
+	metrics.ServiceRequestCounter.With(prometheus.Labels{
+		status.MetricLabelHandler: string(handlerName),
+		status.MetricLabelStatus:  s,
+	}).Inc()
+
+	metrics.ServiceRequestDuration.With(prometheus.Labels{
+		status.MetricLabelHandler: string(handlerName),
+		status.MetricLabelStatus:  s,
+	}).Observe(float64(duration.Nanoseconds()))
```
Comment (smartinov) [inline server/handlerequest.go]: Should be duration in seconds

File: server/handlerequest.go (hunk 4 above)
Comment (loeffert) [reply server/handlerequest.go]: since they are instantiated for both service types: socketserver / webserver

Comment (smartinov) [reply server/handlerequest.go]: what about to add another label "service" and then set socketserver / webserver?
It's more prometheus-like!

**Comment by loeffert** [comment]
PR #14 resolves this PR

**Pull request closed without merging**